### PR TITLE
Allow block-specific settings controls

### PR DESCRIPTION
**Pull request by adamziel**
## Description + Test plan
**The problem**
Let's imagine I'd like to add an additional `MenuItem` to block settings menu in a `core/navigation-link` block. I add the following snippet of code to the `edit.js`:

```js
	<BlockSettingsMenuControls>
		{ () => <MenuItem> My test menu item <MenuItem> }
	</BlockSettingsMenuControls>
```

Nice! This should do the trick. Let's test with a navigation link selected:

Zrzut ekranu 2020-05-27 o 13 54 33

So far so good, maybe we can call it a day? Let's add one more link to be extra sure:

Zrzut ekranu 2020-05-27 o 13 55 59

Uh oh, not good! Let's se what happens when the parent navigation block is selected:

Zrzut ekranu 2020-05-27 o 13 55 39

Wow - that's unexpected.

**Proposed solution**

This behavior occurs because `BlockSettingsMenuControls.Slot` doesn't have any restrictions on block id - it just accepts all the fills from the entire component tree. This plays nicely with MenuItems that are supposed to be in every BlockSettingsMenu such as "duplicate" or "convert", but doesn't support specifying additional controls by the block itself.

This PR introduces a separation between the two - the `BlockSettingsMenuControls` fill now has two modes: block-specific (default) and global. Unless `forAllBlocks` property is set to true, the fill will have an instance-specific name and will not be used by any `BlockSettingsMenu` related to another block (or multi-block selection).

Note that making block-specific mode the default feels more natural, but it also breaks BC - any third party code relying on the default global behavior will no longer work. I'd really like to get a second opinion about this decision before merging.

With this PR applied, adding the snippet of code mentioned above results in a more intuitive outcome:

Zrzut ekranu 2020-05-27 o 14 02 25
Zrzut ekranu 2020-05-27 o 14 02 31
Zrzut ekranu 2020-05-27 o 14 03 10

If this approach gets positive feedback, I'd like to add some tests before merging this PR.

## Types of changes
Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: packages/block-editor/src/components/block-settings-menu-controls/index.js [hunk 1]
```diff
@@ -28,16 +33,41 @@ const BlockSettingsMenuControlsSlot = ( { fillProps, clientIds = null } ) => {
...
+);
+
+export const BlockSettingsMenuControls = ( {
+	forAllBlocks = false,
+	...fillProps
+} ) => {
+	const context = useContext( BlockListBlockContext );
+	if ( ! forAllBlocks ) {
+		// Let's use non-existent ID in case the context is missing
+		const clientId = context?.clientId || 'non-such-id';
```
Comment (adamziel) [inline packages/block-editor/src/components/block-settings-menu-controls/index.js:63]: Do we have any assertions available? I think using a block-specific mode outside of block context should result, at a very least, in a clear error message in the console (if not blowing up entirely).

**Comment by gziolo** [comment]
Did you try you to render this fill conditionally when the block is selected:

```js
isSelected && (
    <BlockSettingsMenuControls>
        <MenuItem> My test menu item <MenuItem>
    </BlockSettingsMenuControls>
)
```

The issue is caused by the fact that you share one menu for all child blocks.

**Comment by adamziel** [comment]
@gziolo this solves the problem in post editor, but there's also the experimental menu management screen where navigator items have an ellipsis menu that's clickable even when the block is not selected:

Zrzut ekranu 2020-05-27 o 15 40 47

In theory we could just select the block when the menu is clicked, but it would shift the focus from the navigator over to the post editor which is bad experience especially for users using primarily their keyboards.

**Comment by adamziel** [comment]
With requirements evolving rapidly, this PR may not be needed in the end. If I'm wrong, let's revisit. See also: https://github.com/WordPress/gutenberg/issues/22089#issuecomment-651177170

**Pull request closed without merging**